### DISCUSSION TITLE
KnexTimeoutError Knex: Timeout acquiring a connection under load

### DISCUSSION THREAD

**Discussion by u12206050** [directus/directus#11051] [Archive]
I am running a load test against the Directus api that I host on Cloudrun, to test if it stands up to expected load.

Unfortunately requests start timing out quite early on and I have a lot of log messages stating `knexTimeoutError`

Seems to happen every time there are more than 10-20 req/s.

I am trying to do an insert, which tiggers an onCreate action which in turn will create another record. I use the same database instance that I get from Directus in all of the requests.

### To Reproduce

Run a directus instance on GCP (free version is enough)

Run a load test against the api inserting a new record. 
Try 10-30req/s

### Errors Shown

`KnexTimeoutError Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?`

### What version of Directus are you using?

9.4.3

### What version of Node.js are you using?

16.4.0

### What database are you using?

mysql 8

### What browser are you using?

Artillery.io

### What operating system are you using?

linux

### How are you deploying Directus?

Cloudrun

**Comment by u12206050** [comment]
So after reading a bit, I found the mention of tarn.js in the docs and realized the min pool size by default is 10.
Cloud run instances have a maximum of 100 allowed connections to sql so I set `DB_POOL_MAX=100` and immediately got much better performance. Will be updating as I get closer to more real life load tests. 
Currently able to do 50req/s but need more.

**Comment by rijkvanzanten** [comment]
@u12206050 Are you running Directus itself horizontally scaled as well? If so, keep in mind that each instance of Directus that's running will use up to the configured MAX connections at the same time. If you're running Directus in for example 8 load balanced containers, you'd have 8 * 100 = 800 connections allocated, where your allowed maximum on the database side is 100. In that case, you want to reduce the maximum number of connections from a single Directus instance to suit the database limitations.

I'll move this to a Q&A discussion, as I believe it's configuration / infrastructure, rather than a bug in Directus itself at this point 🙂

**Comment by u12206050** [comment]
I'm using Google SQL with a MySql instance. I have a very basic instance, and it seems the max amount of connections is 1000. 
I have tested this and reached 299 connections, before my own home network capped out and wouldn't allow any more requests :P At that time I was averaging 50req/s but with varied response times from 200ms to 15,000ms with average being 1,800ms

**Reply by rijkvanzanten** [reply]
Is that using a single instance of Directus? Node is single-threaded, so i would highly recommend looking at horizontal scaling once you reach the 50+ reqs/sec range 🙂

**Reply by u12206050** [reply]
Yes I am running multiple containers. Each container with a single cpu, currently auto scaling between 1 and 80 containers each with max 100 connections and max 160 requests per container before termination.

Am still trying different configurations, as another issue I am having is that any `actions` that are defined to execute after a request is return is not guaranteed to execute since the container only lives for the life span of the request.

**Comment by u12206050** [comment]
So after a LOT of tests I can with confidence say the following:
With low concurrency you can expect lower latency per request (sub 100ms), however a lot more containers need to be started in order to support more requests.

With higher concurrency there is a higher latency (sub 1s). But less containers are started to support the same amount of requests.

My current settings in Cloud Run are as follows:

Max Requests: 10 (concurrency)
Max Containers: 100
Timeout: 60s
env: DB_POOL_MAX: 40

With these settings I get sub 100ms response time on average.

For all my tests I tried to ramp up to sending 50req/s however seemed to keep capping out at 46req/s. According to this [post](https://javascript.plainenglish.io/how-many-requests-can-handle-a-real-world-nodejs-server-side-application-55da7a2f06f3) nodejs with express should be able to handle 15000req/s !!!

So either Directus is very slow, or there is a very unique configuration need that I still haven't found.

If your `DB_POOL_MAX` is too low you will get a lot of Knew timeout errors. If it is too high, and you have a lot of containers you will have a lot of "Too many connections" errors.

Would be great if anyone else has Directus in production with more than 100req/s.

**Reply by rijkvanzanten** [reply]
> Would be great if anyone else has Directus in production with more than 100req/s.

We do. We run them in the 5-8 containers range with bursts of 300req/s and up

**Reply by rijkvanzanten** [reply]
> According to this post nodejs with express should be able to handle 15000req/s !!!

This is a silly comparison. A simple `(req, res) => res.send('hi');` will operate a trillion times faster than any endpoint that actually does something, especially when third party services are introduced, like redis / databases / file stores etc etc.

**Reply by u12206050** [reply]
That is true, although it does give me a baseline of which I hope to handle at least 1% of that. ie. 150req/s

Would be interesting to know the settings ie. vCPU, Memory, Concurrency, DB_POOL* that you use in order to support 300req/s

**Reply by rijkvanzanten** [reply]
> Would be interesting to know the settings ie. vCPU, Memory, Concurrency, DB_POOL* that you use in order to support 300req/s

In my personal uses, I tend to run .75 or 1 vCPU, 1 gig of memory, 1-8 containers (scaling based on usage), and default values for the rest of the pool env vars. That being said, I rely on AWS ECS + Aurora RDS which handles database connections very efficiently, not sure how GCP compares to that ofc

**Reply by fentonfentonfenton** [reply]
@rijkvanzanten heya - what instance size do you use in aurora ?

**Reply by rijkvanzanten** [reply]
Depends on the project I need to power. There's no single right answer for all use cases 👍🏻

**Reply by fentonfentonfenton** [reply]
> Depends on the project I need to power. There's no single right answer for all use cases 👍🏻

I'm with you there - just curious what your aurora instance and pool settings are for ' run .75 or 1 vCPU, 1 gig of memory, 1-8 containers in ECS' (we're trying to do something very similar in production so inspiration is very helpful (we won't take your advice as gospel) )

**Reply by rijkvanzanten** [reply]
> > Depends on the project I need to power. There's no single right answer for all use cases 👍🏻
> 
> 
> 
> I'm with you there - just curious what your aurora instance and pool settings are for ' run .75 or 1 vCPU, 1 gig of memory, 1-8 containers in ECS' (we're trying to do something very similar in production so inspiration is very helpful (we won't take your advice as gospel) )

It still depends.. there's tons of RDS options available, and they all have specific pros and cons. There's a lot more factors at play here than just how many instances of directus are connected, like api traffic, burstability, scaling requirements, cost limitations, backup capabilities, etc etc :)

**Comment by u12206050** [comment]
A big issue when starting to scale seems to be when the connection limit of the POOL_SIZE is reached.

This to me indicates that either Directus is making poor use of each available db connection, or the pool handles new requests poorly.

Other than that a big killer is the event emitter, especially for actions that happen after the request is returned. These action handlers seems to block new incoming requests on the same thread increasing the next requests latency.

So one solution is to not use the Directus event emitter, and use pub/sub or similar. 

btw, I also tried a more simple GET request directly on the Directus api. I managed to reach 300req/s before getting network timeouts, but from around 250req/s the latency increased from 1s to 10s!

**Reply by rijkvanzanten** [reply]
> These action handlers seems to block new incoming requests on the same thread increasing the next requests latency.

Do you have an example we could use to reproduce this?

**Reply by u12206050** [reply]
Sorry forgot to update, after switching on the option in GCP "always allocate CPU" then this issue seemed to happen less, probably because it didn't need to wait for the next request in order to execute.

**Reply by rijkvanzanten** [reply]
I was gonna say. We don't use any synchronous javascript in those api event emitters, so that shouldn't be the case 🙂

**Comment by u12206050** [comment]
Update:

After trying more settings todo with DB_POOL it seems that existing connections are not being reused effectively. I say this because if I destroy existing connections faster than the default setting of 5s then I seem to be able to get more requests through, since there are then available connections.

This is what I have now. 
```
DB_POOL_destroyTimeoutMillis=1000
DB_POOL_idleTimeoutMillis=500
DB_POOL_reapIntervalMillis=500
DB_POOL_createRetryIntervalMillis=500
```


However note, even though I am able to almost reach 100req/s it eventually still ends up throwing ` Knex: Timeout acquiring a connection. The pool is probably full.`

**Comment by u12206050** [comment]
So some improvement setting
`DB_POOL_MIN=0`
But I still eventually hit max connections on the database side, seems to be due to locking of a table that is getting a lot of inserts.

**Reply by joselcvarela** [reply]
Hello @u12206050 
We just tested on a big project and it seems when we are requesting multiple nested fields, we hit a performance issue on database side. Do you think this is what is happening on your side or do you believe it's on insert?

**Reply by u12206050** [reply]
Hi, I do have quite a few nested fields but those queries seem to do fine.
By monitoring the mysql database with `show processlist;` I notice that as soon as there are multiple update commands to the same table (especially the same row) then connections start waiting on each other, but because in the cloud there could be upto 100 containers all doing requests this creates a dead lock scenario.

Unfortunately I have not been able to stably get more than 50req/s

Some things I have done that did improve things a bit:
 
 - Implement a jobs queue, instead of calling emitter, I add a job to the database to be executed later
 - Cache my most common requests in memory even just for a short duration.
 - Batch update and get requests by delaying the execution of code (only within jobs)

**Reply by rijkvanzanten** [reply]
> I notice that as soon as there are multiple update commands to the same table (especially the same row) then connections start waiting on each other

This sounds like a database scaling issue then right? Or are you saying there's a specific request that Directus currently makes that's causing this?

**Reply by u12206050** [reply]
Running the same load test on one machine does produce better results than running the same test on GCP (where multiple containers are trying to execute queries at the same time)

From the mysql log it does seem that there are enough connections, but that existing connections are more likely to be destroyed rather than being reused.

So even though the three highest queries are
`select ip_access from directus_roles where `id` = 'AAAAAAAAAAAAAAA' limit 1`
`select id, role, admin_access, app_access from directus_users left join directus_roles…`
`insert into directus_activity (action, collection, ip, item, user, user_agent) values …`

which all seem directus related, I think the main issue is the amount of connections being destroyed an recreated, leading me to think it is a knex/tarnjs issue.

NOTE: I do have cache enabled, so not sure why permissions & tokens aren't being cached?

**Comment by u12206050** [comment]
Found a possible cause:

When using `filter` for update and create, you should not use the database instance passed by the `HookConfig` instead you should use the database within the context of the filter function since that is actually an instance of a transaction.

Not doing so can lead to a database lock as explained here: [knex pool exhausted](https://github.com/Vincit/objection.js/issues/1137#issuecomment-561149456)

I must admit I have been constantly using the database instance passed in the HookConfig to create a global ItemsService that I "thought" could just be reused everywhere. (just didn't like the idea of creating a new ItemsService everytime)

Still have to run tests...

**Reply by rijkvanzanten** [reply]
That makes total sense. If you have custom logic that hits the database during an existing transaction, knex will have to create a secondary connection to make that query. Reusing the existing database connection (transaction) inside the hook handler function should prevent that overhead 👍🏻

**Reply by NilsBaumgartner1994** [reply]
Facing the same problem, as i was using the connection from the hookConfig.

**Comment by keesvanbemmel** [comment]
Hi @u12206050 I'm interested to see your DB_POOL__* environment variables for your GCP setup. I'm running into similar issues and it seems that your earlier "DB_POOL_MAX" is not available anymore. Any help is greatly appreciated!

**Reply by rijkvanzanten** [reply]
> I'm running into similar issues and it seems that your earlier "DB_POOL_MAX" is not available anymore. 

I think that's a typo, it's `DB_POOL__MAX` instead of `DB_POOL_MAX`, as everything under `DB_POOL_*` is sent straight to the `poolConfig` section of the knex connection :)

**Reply by u12206050** [reply]
Hey, these are my current production env variables regarding DB POOL:

```
DB_POOL_MIN=0
DB_POOL_MAX=60
DB_POOL_destroyTimeoutMillis=2000
DB_POOL_idleTimeoutMillis=1000
DB_POOL_reapIntervalMillis=1000
DB_POOL_createRetryIntervalMillis=100
DB_POOL_acquireTimeoutMillis=10000
DB_POOL_createTimeoutMillis=10000
```

**Reply by keesvanbemmel** [reply]
Shouldn't there be double underscores after POOL?

**Reply by u12206050** [reply]
No, only one

**Reply by keesvanbemmel** [reply]
> > I'm running into similar issues and it seems that your earlier "DB_POOL_MAX" is not available anymore.
> 
> I think that's a typo, it's `DB_POOL__MAX` instead of `DB_POOL_MAX`, as everything under `DB_POOL_*` is sent straight to the `poolConfig` section of the knex connection :)

You mean the typo is in the [docs](https://docs.directus.io/self-hosted/config-options/#database)? @rijkvanzanten 

I just tried it with DB_POOL__MAX and that got in the pool config. Double underscores it seems?

**Reply by rijkvanzanten** [reply]
Pretty sure that's supposed to be two 🤔 

`DB_*` stuff is auto transformed to objects, which in turn means that `DB_POOL_MIN=0` would become `{ poolMin: 0 }` rather than `{ pool: { min: 0 }}` as is required in the knex config

**Reply by u12206050** [reply]
Hmm interesting, so all those variables havn't been doing anything 😅

**Reply by rijkvanzanten** [reply]
To that same effect, you can write `DB_POOL__createTimeoutMillis` as `DB_POOL__CREATE_TIMEOUT_MILLIS` in case you prefer keeping it consistent, but I digress

**Reply by keesvanbemmel** [reply]
Glad that we got that cleared up :)

I now have:
DB_POOL__MIN="10"
DB_POOL__MAX="200"

Let's see how that performs.

Cheers!

**Reply by rijkvanzanten** [reply]
> You mean the typo is in the [docs](https://docs.directus.io/self-hosted/config-options/#database)? @rijkvanzanten

AFAICT the docs list it as `DB_POOL__*`, but (just like on GH) the font makes it a tad difficult to see the difference of _ vs __ 😄

**Reply by JonathanSchndr** [reply]
I have a same error on render.com and here it helped to set the Knex.Config: { acquireConnectionTimeout: 600000 }  - Only my question would be. How can I set this via the Directus ENV?

Update: DB_ACQUIRECONNECTIONTIMEOUT=600000 not working, because the destruction is mapped:
```
const {
    client,
    version,
    searchPath,
    connectionString,
    pool: poolConfig = {},
    ...connectionConfig
  } = getConfigFromEnv('DB_', ['DB_EXCLUDE_TABLES']); 
```

**Reply by rijkvanzanten** [reply]
@JonathanSchndr you should be able to use `DB_ACQUIRE_CONNECTION_TIMEOUT`

**Comment by nbert71** [comment]
Hi,
I encountered this problem, I am working on it and if it helps, on my side, I have some transactions/queries that are taking way too much time leading to stacking transactions, one blocking the other and leading to a database congestion.

In my case, the query that is problematic is in my opinion related to directus_flows retentions :
delete from "directus_activity" where "id" in (select "directus_activity"."id" from "directus_activity" where "timestamp" < $1 and "action" = $2 limit $3)

It activated the retention process in order to delete only flows logs since i run a flow every 30sec, here's my config.

- ACTIVITY_RETENTION=false
- FLOW_LOGS_RETENTION=40d
- RETENTION_BATCH=500
- RETENTION_ENABLED=true
- RETENTION_SCHEDULE=*/5 * * * *
- REVISIONS_RETENTION=false

Maybe i am being too greedy...
I will try to manually delete these lines and not track the logs of the concerned and see if this helps.
